Lms/admin diagnostic performance by skill

anathomical opened the pull request:
## WHAT
- Add a new query to aggregate Diagnostic data for admins by skill (instead of by diagnostic)
- Add a worker and controller to cache and fetch the data from this query

NOTE: This query is currently quite time-consuming (~15 seconds regardless of the size of the dataset being analyzed).  We're intent is to address performance concerns with BigQuery in general next.
## WHY
- We need this data for the "By Skill" part of the Admin Diagnostic reports
- We have an established pattern for these workers and controllers, so we're just doing a bunch of copy/paste work here.  We should follow this project up with a refactor to consolidate the shared logic into some sort of shared code rather than doing all of this duplication
## HOW
- The nature of the data calculation we need to do involves multiple steps of aggregation.  In order to facilitate this, I developed a query with a series of `WITH` statements that feed into each other to successively aggregate the necessary data.  This involved a mild refactor of the parent query in order to support an arbitrary number of `WITH` statements
- - This also involved adding support for a new "type" of aggregation in the parent query to handle the calculation of percentages
- Copy and paste from the reference worker and controller (in this case, the one for the Admin Diagnostic Overview), and modify as appropriate

### Notion Card Links
https://www.notion.so/quill/Admin-Diagnostic-Growth-Report-Growth-Per-Skill-1cdc878fb976477ea40089ddef6711ca?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | (N/A or Yes)

anathomical committed:
WIP finalizing logic for proficiency calculation

anathomical committed:
Finish working out improvement calculations

anathomical committed:
Add controller, route, and spec for Skills controller

anathomical committed:
Add Diagnostic Skills worker

anathomical committed:
Add a test case for new percentage_aggregate functionality

anathomical committed:
Create a spec for Diagnostic By Skill query

This has a minimum number of cases because this query is so time-consuming at the moment.  We should add more cases once we resolve this

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/admin-diagnostic-performance-by-skill

anathomical committed:
Update controller spec to work

brendanshean commented on services/QuillLMS/app/controllers/admin_diagnostic_skills_controller.rb:
```diff
@@ -0,0 +1,116 @@
+
+    return render json: { error: 'diagnostic_id param is required' }, status: 400 unless school_ids_param_valid?
+
+    return render json: { error: 'unrecognized query type for this endpoint' }, status: 400 unless WORKERS_FOR_ACTIONS[action_name]::QUERIES.keys.include?(@query)
```
For consistency, what about a boolean method for this one too:

```return render json: { error: 'unrecognized query type for this endpoint' }, status: 400 unless valid_query_type?```

brendanshean commented on services/QuillLMS/spec/queries/admin_diagnostic_reports/diagnostic_performance_by_skill_query_spec.rb:
```diff
@@ -0,0 +1,149 @@
+
+        expect(result[:aggregate_rows].length).to eq(classrooms.map(&:grade).uniq.length)
+        expect(result[:skill_name]).to eq(skill_group.name)
+        expect(result[:pre_score]).to eq(all_pre_concept_results.select { |cr| cr.correct }.length.to_f / all_pre_concept_results.length * 100)
```
I think the following four `select { |cr| cr.correct }` can be rewritten `select(&:correct)`

brendanshean commented on services/QuillLMS/spec/queries/admin_diagnostic_reports/diagnostic_performance_by_skill_query_spec.rb:
```diff
@@ -0,0 +1,149 @@
+        expect(result[:post_correct_total]).to eq(all_post_concept_results.select { |cr| cr.correct }.length)
+        expect(result[:post_total_questions]).to eq(all_post_concept_results.length)
+
+        # The setup for these tests gives users 50% correct rates on their pre-diagnostic, and 100% correct rates on their post-diagnostic, which correspond to "imrpoved proficiency"
```
*improved

brendanshean reviewed (commented):
Great job figuring out these complex aggregations.  I also think the underlying refactor make this more readable.

dandrabik commented on services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_query.rb:
```diff
@@ -42,16 +43,33 @@ def run_query
+      end.join(",\n")
+    end
+
+    def with_queries
```
One thing that would be nice is a way to generate the SQL consistently across these classes, for debugging/optimizing purposes. For the Snapshot queries, that method was `query` and for these it's `union_query` which took me a second to find. We should consider making this `query` for all, or adding an additional method like `to_sql` that can be used consistently.

dandrabik commented on services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_performance_by_skill_query.rb:
```diff
@@ -0,0 +1,233 @@
+      SQL
+    end
+
+    def with_queries
```
Yeah, this is a huge query, the 15 seconds makes more sense now. We'll have to investigate ways of doing this differently (multiple queries, pre-processing data to get it in a better shape, etc.)

dandrabik reviewed (approved):
Overall the setup looks good. Now we just have to dig into the query performance. Thanks for putting this together.

anathomical committed:
Make tweaks based on Brendan's feedback

anathomical committed:
Tweak method names for `query` in the query hierarchy

So that the `query` method always returns the string for the final query that will be sent to BigQuery so that we have a consistent method for seeing the final construction
